Dia 4 - Dataset Movies

In [4]:
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

import os
for dirname, _, filenames in os.walk('C:/Users/patri/OneDrive/Documentos/Códigos/Python/7-days-of-code-CD/Data/ml-100k'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current ses

C:/Users/patri/OneDrive/Documentos/Códigos/Python/7-days-of-code-CD/Data/ml-100k\allbut.pl
C:/Users/patri/OneDrive/Documentos/Códigos/Python/7-days-of-code-CD/Data/ml-100k\mku.sh
C:/Users/patri/OneDrive/Documentos/Códigos/Python/7-days-of-code-CD/Data/ml-100k\README
C:/Users/patri/OneDrive/Documentos/Códigos/Python/7-days-of-code-CD/Data/ml-100k\u.data
C:/Users/patri/OneDrive/Documentos/Códigos/Python/7-days-of-code-CD/Data/ml-100k\u.genre
C:/Users/patri/OneDrive/Documentos/Códigos/Python/7-days-of-code-CD/Data/ml-100k\u.info
C:/Users/patri/OneDrive/Documentos/Códigos/Python/7-days-of-code-CD/Data/ml-100k\u.item
C:/Users/patri/OneDrive/Documentos/Códigos/Python/7-days-of-code-CD/Data/ml-100k\u.occupation
C:/Users/patri/OneDrive/Documentos/Códigos/Python/7-days-of-code-CD/Data/ml-100k\u.user
C:/Users/patri/OneDrive/Documentos/Códigos/Python/7-days-of-code-CD/Data/ml-100k\u1.base
C:/Users/patri/OneDrive/Documentos/Códigos/Python/7-days-of-code-CD/Data/ml-100k\u1.test
C:/Users/patri/OneDr

In [9]:
data_filmes = pd.read_csv('C:/Users/patri/OneDrive/Documentos/Códigos/Python/7-days-of-code-CD/Data/ml-100k/u.item', sep='|', encoding='latin-1', header=None, index_col=False)
data_filmes.head()

0                  1            2   3    
0   1   Toy Story (1995)  01-Jan-1995 NaN  \
1   2   GoldenEye (1995)  01-Jan-1995 NaN   
2   3  Four Rooms (1995)  01-Jan-1995 NaN   
3   4  Get Shorty (1995)  01-Jan-1995 NaN   
4   5     Copycat (1995)  01-Jan-1995 NaN   

                                                  4   5   6   7   8   9   ...   
0  http://us.imdb.com/M/title-exact?Toy%20Story%2...   0   0   0   1   1  ...  \
1  http://us.imdb.com/M/title-exact?GoldenEye%20(...   0   1   1   0   0  ...   
2  http://us.imdb.com/M/title-exact?Four%20Rooms%...   0   0   0   0   0  ...   
3  http://us.imdb.com/M/title-exact?Get%20Shorty%...   0   1   0   0   0  ...   
4  http://us.imdb.com/M/title-exact?Copycat%20(1995)   0   0   0   0   0  ...   

   14  15  16  17  18  19  20  21  22  23  
0   0   0   0   0   0   0   0   0   0   0  
1   0   0   0   0   0   0   0   1   0   0  
2   0   0   0   0   0   0   0   1   0   0  
3   0   0   0   0   0   0   0   0   0   0  
4   0   0   0   0   0   0   0   1   0   0  

[5 rows x 24 columns]

In [16]:
data_filmes.columns = ['movie_id', 'movie_title', 'release_date', 'video_release_date', 'IMDb_URL', 'unknown',
                 'Action', 'Adventure', 'Animation', 'Children', 'Comedy', 'Crime', 'Documentary', 
                 'Drama', 'Fantasy', 'FilmNoir', 'Horror', 'Musical', 'Mystery', 'Romance', 'SciFi', 
                 'Thriller', 'War', 'Western']
data_filmes.head

<bound method NDFrame.head of       movie_id                                movie_title release_date   
0            1                           Toy Story (1995)  01-Jan-1995  \
1            2                           GoldenEye (1995)  01-Jan-1995   
2            3                          Four Rooms (1995)  01-Jan-1995   
3            4                          Get Shorty (1995)  01-Jan-1995   
4            5                             Copycat (1995)  01-Jan-1995   
...        ...                                        ...          ...   
1677      1678                          Mat' i syn (1997)  06-Feb-1998   
1678      1679                           B. Monkey (1998)  06-Feb-1998   
1679      1680                       Sliding Doors (1998)  01-Jan-1998   
1680      1681                        You So Crazy (1994)  01-Jan-1994   
1681      1682  Scream of Stone (Schrei aus Stein) (1991)  08-Mar-1996   

      video_release_date                                           IMDb_URL   
0 

In [12]:
data_filmes.shape

(1682, 24)

In [13]:
data_filmes.info

<bound method DataFrame.info of         0                                          1            2   3    
0        1                           Toy Story (1995)  01-Jan-1995 NaN  \
1        2                           GoldenEye (1995)  01-Jan-1995 NaN   
2        3                          Four Rooms (1995)  01-Jan-1995 NaN   
3        4                          Get Shorty (1995)  01-Jan-1995 NaN   
4        5                             Copycat (1995)  01-Jan-1995 NaN   
...    ...                                        ...          ...  ..   
1677  1678                          Mat' i syn (1997)  06-Feb-1998 NaN   
1678  1679                           B. Monkey (1998)  06-Feb-1998 NaN   
1679  1680                       Sliding Doors (1998)  01-Jan-1998 NaN   
1680  1681                        You So Crazy (1994)  01-Jan-1994 NaN   
1681  1682  Scream of Stone (Schrei aus Stein) (1991)  08-Mar-1996 NaN   

                                                     4   5   6   7   8   9    


In [17]:
data_filmes.video_release_date.unique()

array([nan])

In [18]:
data_filmes.drop(columns= ['video_release_date'], inplace = True)
data_filmes.head()

movie_id        movie_title release_date   
0         1   Toy Story (1995)  01-Jan-1995  \
1         2   GoldenEye (1995)  01-Jan-1995   
2         3  Four Rooms (1995)  01-Jan-1995   
3         4  Get Shorty (1995)  01-Jan-1995   
4         5     Copycat (1995)  01-Jan-1995   

                                            IMDb_URL  unknown  Action   
0  http://us.imdb.com/M/title-exact?Toy%20Story%2...        0       0  \
1  http://us.imdb.com/M/title-exact?GoldenEye%20(...        0       1   
2  http://us.imdb.com/M/title-exact?Four%20Rooms%...        0       0   
3  http://us.imdb.com/M/title-exact?Get%20Shorty%...        0       1   
4  http://us.imdb.com/M/title-exact?Copycat%20(1995)        0       0   

   Adventure  Animation  Children  Comedy  ...  Fantasy  FilmNoir  Horror   
0          0          1         1       1  ...        0         0       0  \
1          1          0         0       0  ...        0         0       0   
2          0          0         0       0  ...        0         0       0   
3          0          0         0       1  ...        0         0       0   
4          0          0         0       0  ...        0         0       0   

   Musical  Mystery  Romance  SciFi  Thriller  War  Western  
0        0        0        0      0         0    0        0  
1        0        0        0      0         1    0        0  
2        0        0        0      0         1    0        0  
3        0        0        0      0         0    0        0  
4        0        0        0      0         1    0        0  

[5 rows x 23 columns]

In [19]:
notas = pd.read_csv('C:/Users/patri/OneDrive/Documentos/Códigos/Python/7-days-of-code-CD/Data/ml-100k/u.data', encoding = 'latin-1', sep = "\t", header = None, index_col = False)
notas.columns = ['user_id', 'item_id', 'rating', 'timestamp']
notas.head()

user_id  item_id  rating  timestamp
0      196      242       3  881250949
1      186      302       3  891717742
2       22      377       1  878887116
3      244       51       2  880606923
4      166      346       1  886397596

In [20]:
notas.shape

(100000, 4)

In [21]:
notas.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 4 columns):
 #   Column     Non-Null Count   Dtype
---  ------     --------------   -----
 0   user_id    100000 non-null  int64
 1   item_id    100000 non-null  int64
 2   rating     100000 non-null  int64
 3   timestamp  100000 non-null  int64
dtypes: int64(4)
memory usage: 3.1 MB


In [22]:
notas.describe()

user_id        item_id         rating     timestamp
count  100000.00000  100000.000000  100000.000000  1.000000e+05
mean      462.48475     425.530130       3.529860  8.835289e+08
std       266.61442     330.798356       1.125674  5.343856e+06
min         1.00000       1.000000       1.000000  8.747247e+08
25%       254.00000     175.000000       3.000000  8.794487e+08
50%       447.00000     322.000000       4.000000  8.828269e+08
75%       682.00000     631.000000       4.000000  8.882600e+08
max       943.00000    1682.000000       5.000000  8.932866e+08

In [23]:
# Escolhendo o ID dos filmes como index
data_filmes = data_filmes.set_index('movie_id')

# Selecionando um filme através do seu ID (escolhendo pelo horário que estou digitando esta linha)
data_filmes.loc[1143]

movie_title                              Hard Eight (1996)
release_date                                   28-Feb-1997
IMDb_URL        http://us.imdb.com/Title?Hard+Eight+(1996)
unknown                                                  0
Action                                                   0
Adventure                                                0
Animation                                                0
Children                                                 0
Comedy                                                   0
Crime                                                    1
Documentary                                              0
Drama                                                    0
Fantasy                                                  0
FilmNoir                                                 0
Horror                                                   0
Musical                                                  0
Mystery                                                 

Realizando verificação do total de votos de cada filme

In [24]:
totally_of_votes = notas['item_id'].value_counts()
totally_of_votes.head()

item_id
50     583
258    509
100    508
181    507
294    485
Name: count, dtype: int64

In [25]:
data_filmes['Total_de_Votos'] = totally_of_votes
data_filmes.head()

movie_title release_date   
movie_id                                   
1          Toy Story (1995)  01-Jan-1995  \
2          GoldenEye (1995)  01-Jan-1995   
3         Four Rooms (1995)  01-Jan-1995   
4         Get Shorty (1995)  01-Jan-1995   
5            Copycat (1995)  01-Jan-1995   

                                                   IMDb_URL  unknown  Action   
movie_id                                                                       
1         http://us.imdb.com/M/title-exact?Toy%20Story%2...        0       0  \
2         http://us.imdb.com/M/title-exact?GoldenEye%20(...        0       1   
3         http://us.imdb.com/M/title-exact?Four%20Rooms%...        0       0   
4         http://us.imdb.com/M/title-exact?Get%20Shorty%...        0       1   
5         http://us.imdb.com/M/title-exact?Copycat%20(1995)        0       0   

          Adventure  Animation  Children  Comedy  Crime  ...  FilmNoir   
movie_id                                                 ...             
1                 0          1         1       1      0  ...         0  \
2                 1          0         0       0      0  ...         0   
3                 0          0         0       0      0  ...         0   
4                 0          0         0       1      0  ...         0   
5                 0          0         0       0      1  ...         0   

          Horror  Musical  Mystery  Romance  SciFi  Thriller  War  Western   
movie_id                                                                     
1              0        0        0        0      0         0    0        0  \
2              0        0        0        0      0         1    0        0   
3              0        0        0        0      0         1    0        0   
4              0        0        0        0      0         0    0        0   
5              0        0        0        0      0         1    0        0   

          Total_de_Votos  
movie_id                  
1                    452  
2                    131  
3                     90  
4                    209  
5                     86  

[5 rows x 23 columns]

In [27]:
data_filmes.sort_values('Total_de_Votos', ascending = False)


movie_title release_date   
movie_id                                                                   
50                                         Star Wars (1977)  01-Jan-1977  \
258                                          Contact (1997)  11-Jul-1997   
100                                            Fargo (1996)  14-Feb-1997   
181                               Return of the Jedi (1983)  14-Mar-1997   
294                                        Liar Liar (1997)  21-Mar-1997   
...                                                     ...          ...   
1576                         Hungarian Fairy Tale, A (1987)  01-Jan-1987   
1577      Death in the Garden (Mort en ce jardin, La) (1...  01-Jan-1956   
1348                             Every Other Weekend (1990)  01-Jan-1990   
1579                                     Baton Rouge (1988)  01-Jan-1988   
1682              Scream of Stone (Schrei aus Stein) (1991)  08-Mar-1996   

                                                   IMDb_URL  unknown  Action   
movie_id                                                                       
50        http://us.imdb.com/M/title-exact?Star%20Wars%2...        0       1  \
258               http://us.imdb.com/Title?Contact+(1997/I)        0       0   
100         http://us.imdb.com/M/title-exact?Fargo%20(1996)        0       0   
181       http://us.imdb.com/M/title-exact?Return%20of%2...        0       1   
294               http://us.imdb.com/Title?Liar+Liar+(1997)        0       0   
...                                                     ...      ...     ...   
1576      http://us.imdb.com/M/title-exact?Hol%20volt,%2...        0       0   
1577      http://us.imdb.com/Title?Mort+en+ce+jardin,+La...        0       0   
1348      http://us.imdb.com/Title?Un+week-end+sur+deux+...        0       0   
1579          http://us.imdb.com/Title?B%E2ton+rouge+(1988)        0       0   
1682      http://us.imdb.com/M/title-exact?Schrei%20aus%...        0       0   

          Adventure  Animation  Children  Comedy  Crime  ...  FilmNoir   
movie_id                                                 ...             
50                1          0         0       0      0  ...         0  \
258               0          0         0       0      0  ...         0   
100               0          0         0       0      1  ...         0   
181               1          0         0       0      0  ...         0   
294               0          0         0       1      0  ...         0   
...             ...        ...       ...     ...    ...  ...       ...   
1576              0          0         0       0      0  ...         0   
1577              0          0         0       0      0  ...         0   
1348              0          0         0       0      0  ...         0   
1579              0          0         0       0      0  ...         0   
1682              0          0         0       0      0  ...         0   

          Horror  Musical  Mystery  Romance  SciFi  Thriller  War  Western   
movie_id                                                                     
50             0        0        0        1      1         0    1        0  \
258            0        0        0        0      1         0    0        0   
100            0        0        0        0      0         1    0        0   
181            0        0        0        1      1         0    1        0   
294            0        0        0        0      0         0    0        0   
...          ...      ...      ...      ...    ...       ...  ...      ...   
1576           0        0        0        0      0         0    0        0   
1577           0        0        0        0      0         0    0        0   
1348           0        0        0        0      0         0    0        0   
1579           0        0        0        0      0         1    0        0   
1682           0        0        0        0      0         0    0        0   

          Total_de_Votos  
movie_id     

In [28]:
notas_medias = notas.groupby("item_id").mean()['rating']
notas_medias.head()

item_id
1    3.878319
2    3.206107
3    3.033333
4    3.550239
5    3.302326
Name: rating, dtype: float64

In [29]:
data_filmes['Nota_Media'] = notas_medias
data_filmes.head()

movie_title release_date   
movie_id                                   
1          Toy Story (1995)  01-Jan-1995  \
2          GoldenEye (1995)  01-Jan-1995   
3         Four Rooms (1995)  01-Jan-1995   
4         Get Shorty (1995)  01-Jan-1995   
5            Copycat (1995)  01-Jan-1995   

                                                   IMDb_URL  unknown  Action   
movie_id                                                                       
1         http://us.imdb.com/M/title-exact?Toy%20Story%2...        0       0  \
2         http://us.imdb.com/M/title-exact?GoldenEye%20(...        0       1   
3         http://us.imdb.com/M/title-exact?Four%20Rooms%...        0       0   
4         http://us.imdb.com/M/title-exact?Get%20Shorty%...        0       1   
5         http://us.imdb.com/M/title-exact?Copycat%20(1995)        0       0   

          Adventure  Animation  Children  Comedy  Crime  ...  Horror  Musical   
movie_id                                                 ...                    
1                 0          1         1       1      0  ...       0        0  \
2                 1          0         0       0      0  ...       0        0   
3                 0          0         0       0      0  ...       0        0   
4                 0          0         0       1      0  ...       0        0   
5                 0          0         0       0      1  ...       0        0   

          Mystery  Romance  SciFi  Thriller  War  Western  Total_de_Votos   
movie_id                                                                    
1               0        0      0         0    0        0             452  \
2               0        0      0         1    0        0             131   
3               0        0      0         1    0        0              90   
4               0        0      0         0    0        0             209   
5               0        0      0         1    0        0              86   

          Nota_Media  
movie_id              
1           3.878319  
2           3.206107  
3           3.033333  
4           3.550239  
5           3.302326  

[5 rows x 24 columns]

In [30]:
data_filmes.sort_values("Nota_Media", ascending= False).head(15)

movie_title release_date   
movie_id                                                                   
814                           Great Day in Harlem, A (1994)  01-Jan-1994  \
1599                          Someone Else's America (1995)  10-May-1996   
1201             Marlene Dietrich: Shadow and Light (1996)   02-Apr-1996   
1122                         They Made Me a Criminal (1939)  01-Jan-1939   
1653      Entertaining Angels: The Dorothy Day Story (1996)  27-Sep-1996   
1293                                        Star Kid (1997)  16-Jan-1998   
1500                              Santa with Muscles (1996)  08-Nov-1996   
1189                                     Prefontaine (1997)  24-Jan-1997   
1536                                   Aiqing wansui (1994)  22-Jul-1996   
1467                   Saint of Fort Washington, The (1993)  01-Jan-1993   
1449                                 Pather Panchali (1955)  22-Mar-1996   
119                  Maya Lin: A Strong Clear Vision (1994)  01-Jan-1994   
1398                                            Anna (1996)  13-Nov-1996   
1642                               Some Mother's Son (1996)  27-Dec-1996   
1594                                         Everest (1998)  10-Mar-1998   

                                                   IMDb_URL  unknown  Action   
movie_id                                                                       
814       http://us.imdb.com/M/title-exact?Great%20Day%2...        0       0  \
1599      http://us.imdb.com/M/title-exact?Someone%20Els...        0       0   
1201      http://us.imdb.com/M/title-exact?Marlene%20Die...        0       0   
1122      http://us.imdb.com/M/title-exact?They%20Made%2...        0       0   
1653      http://us.imdb.com/M/title-exact?Entertaining%...        0       0   
1293      http://us.imdb.com/M/title-exact?imdb-title-12...        0       0   
1500      http://us.imdb.com/M/title-exact?Santa%20with%...        0       0   
1189      http://us.imdb.com/M/title-exact?Prefontaine%2...        0       0   
1536      http://us.imdb.com/M/title-exact?Aiqing%20Wans...        0       0   
1467      http://us.imdb.com/M/title-exact?Saint%20of%20...        0       0   
1449      http://us.imdb.com/M/title-exact?Pather%20Panc...        0       0   
119       http://us.imdb.com/M/title-exact?Maya%20Lin:%2...        0       0   
1398         http://us.imdb.com/M/title-exact?Anna%20(1996)        0       0   
1642      http://us.imdb.com/M/title-exact?Some%20Mother...        0       0   
1594                http://us.imdb.com/Title?Everest+(1998)        0       0   

          Adventure  Animation  Children  Comedy  Crime  ...  Horror  Musical   
movie_id                                                 ...                    
814               0          0         0       0      0  ...       0        0  \
1599              0          0         0       0      0  ...       0        0   
1201              0          0         0       0      0  ...       0        0   
1122              0          0         0       0      1  ...       0        0   
1653              0          0         0       0      0  ...       0        0   
1293              1          0         1       0      0  ...       0        0   
1500              0          0         0       1      0  ...       0        0   
1189              0          0         0       0      0  ...       0        0   
1536              0          0         0       0      0  ...       0        0   
1467              0          0         0       0      0  ...       0        0   
1449              0          0         0       0      0  ...       0        0   
119               0          0         0       0      0  ...       0        0   
1398              0          0         0       0      0  ...       0        0   
1642              0          0         0       0      0  ...       0        0   
1594              0          0         0       0      0  ...       0        0   

          Myster

In [31]:
data_filmes.query('Total_de_Votos >= 10').sort_values('Nota_Media', ascending= False)

movie_title release_date   
movie_id                                                                   
408                                   Close Shave, A (1995)  28-Apr-1996  \
318                                 Schindler's List (1993)  01-Jan-1993   
169                              Wrong Trousers, The (1993)  01-Jan-1993   
483                                       Casablanca (1942)  01-Jan-1942   
114       Wallace & Gromit: The Best of Aardman Animatio...  05-Apr-1996   
...                                                     ...          ...   
758               Lawnmower Man 2: Beyond Cyberspace (1996)  01-Jan-1996   
1087                                    Bloodsport 2 (1995)  01-Mar-1996   
440                    Amityville II: The Possession (1982)  01-Jan-1982   
669                                       Body Parts (1991)  01-Jan-1991   
424              Children of the Corn: The Gathering (1996)  01-Jan-1996   

                                                   IMDb_URL  unknown  Action   
movie_id                                                                       
408       http://us.imdb.com/M/title-exact?Close%20Shave...        0       0  \
318       http://us.imdb.com/M/title-exact?Schindler's%2...        0       0   
169       http://us.imdb.com/M/title-exact?Wrong%20Trous...        0       0   
483       http://us.imdb.com/M/title-exact?Casablanca%20...        0       0   
114       http://us.imdb.com/Title?Wallace+%26+Gromit%3A...        0       0   
...                                                     ...      ...     ...   
758       http://us.imdb.com/M/title-exact?Lawnmower%20M...        0       0   
1087      http://us.imdb.com/M/title-exact?Bloodsport%20...        0       1   
440       http://us.imdb.com/M/title-exact?Amityville%20...        0       0   
669       http://us.imdb.com/M/title-exact?Body%20Parts%...        0       0   
424       http://us.imdb.com/M/title-exact?Children%20of...        0       0   

          Adventure  Animation  Children  Comedy  Crime  ...  Horror  Musical   
movie_id                                                 ...                    
408               0          1         0       1      0  ...       0        0  \
318               0          0         0       0      0  ...       0        0   
169               0          1         0       1      0  ...       0        0   
483               0          0         0       0      0  ...       0        0   
114               0          1         0       0      0  ...       0        0   
...             ...        ...       ...     ...    ...  ...     ...      ...   
758               0          0         0       0      0  ...       0        0   
1087              0          0         0       0      0  ...       0        0   
440               0          0         0       0      0  ...       1        0   
669               0          0         0       0      0  ...       1        0   
424               0          0         0       0      0  ...       1        0   

          Mystery  Romance  SciFi  Thriller  War  Western  Total_de_Votos   
movie_id                                                                    
408             0        0      0         1    0        0             112  \
318             0        0      0         0    1        0             298   
169             0        0      0         0    0        0             118   
483             0        1      0         0    1        0             243   
114             0        0      0         0    0        0              67   
...           ...      ...    ...       ...  ...      ...             ...   
758             0        0      1         1    0        0              21   
1087            0        0      0         0    0        0              10   
440             0        0      0         0    0        0              14   
669             0        0      0         0    0        0              13   
424             0        0      0       

Fazer a recomendação somente pela média das notas não é uma boa saída, porque filmes desconhecidos podem receber poucas avaliações que o classificam melhor do que os outros e isso faz com que a nota não represente de forma adequada sua avaliação.

In [33]:
data_filmes.Total_de_Votos.min()

1

In [34]:
data_filmes.Total_de_Votos.max()

583

Os valores mínimo e máximos que o filme receberam obtivemos nas últimas duas linhas, vamos trabalhar com filmes que obtiveram até 100 avaliações

In [35]:
filmes_com_mais_de_100_votos = data_filmes.query('Total_de_Votos >= 50')
filmes_com_mais_de_100_votos.sort_values('Nota_Media', ascending = False)

movie_title release_date   
movie_id                                                                   
408                                   Close Shave, A (1995)  28-Apr-1996  \
318                                 Schindler's List (1993)  01-Jan-1993   
169                              Wrong Trousers, The (1993)  01-Jan-1993   
483                                       Casablanca (1942)  01-Jan-1942   
114       Wallace & Gromit: The Best of Aardman Animatio...  05-Apr-1996   
...                                                     ...          ...   
122                                   Cable Guy, The (1996)  14-Jun-1996   
988                    Beautician and the Beast, The (1997)  07-Feb-1997   
120                                       Striptease (1996)  28-Jun-1996   
687                                    McHale's Navy (1997)  18-Apr-1997   
931                        Island of Dr. Moreau, The (1996)  23-Aug-1996   

                                                   IMDb_URL  unknown  Action   
movie_id                                                                       
408       http://us.imdb.com/M/title-exact?Close%20Shave...        0       0  \
318       http://us.imdb.com/M/title-exact?Schindler's%2...        0       0   
169       http://us.imdb.com/M/title-exact?Wrong%20Trous...        0       0   
483       http://us.imdb.com/M/title-exact?Casablanca%20...        0       0   
114       http://us.imdb.com/Title?Wallace+%26+Gromit%3A...        0       0   
...                                                     ...      ...     ...   
122       http://us.imdb.com/M/title-exact?Cable%20Guy,%...        0       0   
988       http://us.imdb.com/M/title-exact?Beautician%20...        0       0   
120       http://us.imdb.com/M/title-exact?Striptease%20...        0       0   
687       http://us.imdb.com/M/title-exact?McHale's%20Na...        0       0   
931       http://us.imdb.com/M/title-exact?Island%20of%2...        0       0   

          Adventure  Animation  Children  Comedy  Crime  ...  Horror  Musical   
movie_id                                                 ...                    
408               0          1         0       1      0  ...       0        0  \
318               0          0         0       0      0  ...       0        0   
169               0          1         0       1      0  ...       0        0   
483               0          0         0       0      0  ...       0        0   
114               0          1         0       0      0  ...       0        0   
...             ...        ...       ...     ...    ...  ...     ...      ...   
122               0          0         0       1      0  ...       0        0   
988               0          0         0       1      0  ...       0        0   
120               0          0         0       1      1  ...       0        0   
687               0          0         0       1      0  ...       0        0   
931               0          0         0       0      0  ...       0        0   

          Mystery  Romance  SciFi  Thriller  War  Western  Total_de_Votos   
movie_id                                                                    
408             0        0      0         1    0        0             112  \
318             0        0      0         0    1        0             298   
169             0        0      0         0    0        0             118   
483             0        1      0         0    1        0             243   
114             0        0      0         0    0        0              67   
...           ...      ...    ...       ...  ...      ...             ...   
122             0        0      0         0    0        0             106   
988             0        1      0         0    0        0              86   
120             0        0      0         0    0        0              67   
687             0        0      0         0    1        0              69   
931             0        0      1       

Fazendo um sistema de recomendação utilizando os gêneros dos filmes:

In [36]:
filmes_com_mais_de_100_votos.sample(10)

movie_title release_date   
movie_id                                                        
20                     Angels and Insects (1995)  01-Jan-1995  \
186                   Blues Brothers, The (1980)  01-Jan-1980   
303                           Ulee's Gold (1997)  01-Jan-1997   
168       Monty Python and the Holy Grail (1974)  01-Jan-1974   
255              My Best Friend's Wedding (1997)  20-Jun-1997   
658                    Pump Up the Volume (1990)  01-Jan-1990   
636                  Escape from New York (1981)  01-Jan-1981   
825                          Arrival, The (1996)  31-May-1996   
82                          Jurassic Park (1993)  01-Jan-1993   
491         Adventures of Robin Hood, The (1938)  01-Jan-1938   

                                                   IMDb_URL  unknown  Action   
movie_id                                                                       
20        http://us.imdb.com/M/title-exact?Angels%20and%...        0       0  \
186       http://us.imdb.com/M/title-exact?Blues%20Broth...        0       1   
303       http://us.imdb.com/M/title-exact?Ulee%27s+Gold...        0       0   
168       http://us.imdb.com/M/title-exact?Monty%20Pytho...        0       0   
255       http://us.imdb.com/M/title-exact?My+Best+Frien...        0       0   
658       http://us.imdb.com/M/title-exact?Pump%20Up%20t...        0       0   
636       http://us.imdb.com/M/title-exact?Escape%20from...        0       1   
825       http://us.imdb.com/M/title-exact?Arrival,%20Th...        0       1   
82        http://us.imdb.com/M/title-exact?Jurassic%20Pa...        0       1   
491       http://us.imdb.com/M/title-exact?Adventures%20...        0       1   

          Adventure  Animation  Children  Comedy  Crime  ...  Horror  Musical   
movie_id                                                 ...                    
20                0          0         0       0      0  ...       0        0  \
186               0          0         0       1      0  ...       0        1   
303               0          0         0       0      0  ...       0        0   
168               0          0         0       1      0  ...       0        0   
255               0          0         0       1      0  ...       0        0   
658               0          0         0       0      0  ...       0        0   
636               1          0         0       0      0  ...       0        0   
825               0          0         0       0      0  ...       0        0   
82                1          0         0       0      0  ...       0        0   
491               1          0         0       0      0  ...       0        0   

          Mystery  Romance  SciFi  Thriller  War  Western  Total_de_Votos   
movie_id                                                                    
20              0        1      0         0    0        0              72  \
186             0        0      0         0    0        0             251   
303             0        0      0         0    0        0             134   
168             0        0      0         0    0        0             316   
255             0        1      0         0    0        0             172   
658             0        0      0         0    0        0              79   
636             0        0      1         1    0        0              91   
825             0        0      1         1    0        0              83   
82              0        0      1         0    0        0             261   
491             0        0      0         0    0        0              67   

          Nota_Media  
movie_id              
20          3.416667  
186         3.836653  
303         3.686567  
168         4.066456  
255         3.360465  
658         3.417722  
636         3.241758  
825         3.036145  
82          3.720307  
491         3.791045  

[10 rows x 24 columns]

Supondo que assistimos os filmes da lista acima e queiramos assistir um filme de ação e drama.

In [38]:
assistido = [20, 186, 303, 168, 255, 658, 636, 825, 82, 491]
filmes_com_mais_de_100_votos.query('Action == 1 and Drama == 1')

movie_title release_date   
movie_id                                                 
4                       Get Shorty (1995)  01-Jan-1995  \
22                      Braveheart (1995)  16-Feb-1996   
28                       Apollo 13 (1995)  01-Jan-1995   
54                        Outbreak (1995)  01-Jan-1995   
127                 Godfather, The (1972)  01-Jan-1972   
172       Empire Strikes Back, The (1980)  01-Jan-1980   
187        Godfather: Part II, The (1974)  01-Jan-1974   
188              Full Metal Jacket (1987)  01-Jan-1987   
207             Cyrano de Bergerac (1990)  01-Jan-1990   
245               Devil's Own, The (1997)  26-Mar-1997   
313                        Titanic (1997)  01-Jan-1997   
326                      G.I. Jane (1997)  01-Jan-1997   
403                         Batman (1989)  01-Jan-1989   
510         Magnificent Seven, The (1954)  01-Jan-1954   
515                      Boot, Das (1981)  04-Apr-1997   
526                        Ben-Hur (1959)  01-Jan-1959   
631               Crying Game, The (1992)  01-Jan-1992   
651                          Glory (1989)  01-Jan-1989   
686               Perfect World, A (1993)  01-Jan-1993   
720                   First Knight (1995)  01-Jan-1995   
806              Menace II Society (1993)  01-Jan-1993   
840              Last Man Standing (1996)  20-Sep-1996   
855                           Diva (1981)  01-Jan-1981   

                                                   IMDb_URL  unknown  Action   
movie_id                                                                       
4         http://us.imdb.com/M/title-exact?Get%20Shorty%...        0       1  \
22        http://us.imdb.com/M/title-exact?Braveheart%20...        0       1   
28        http://us.imdb.com/M/title-exact?Apollo%2013%2...        0       1   
54        http://us.imdb.com/M/title-exact?Outbreak%20(1...        0       1   
127       http://us.imdb.com/M/title-exact?Godfather,%20...        0       1   
172       http://us.imdb.com/M/title-exact?Empire%20Stri...        0       1   
187       http://us.imdb.com/M/title-exact?Godfather:%20...        0       1   
188       http://us.imdb.com/M/title-exact?Full%20Metal%...        0       1   
207       http://us.imdb.com/M/title-exact?Cyrano%20de%2...        0       1   
245       http://us.imdb.com/M/title-exact?Devil%27s%20O...        0       1   
313       http://us.imdb.com/M/title-exact?imdb-title-12...        0       1   
326       http://us.imdb.com/M/title-exact?G%2EI%2E+Jane...        0       1   
403        http://us.imdb.com/M/title-exact?Batman%20(1989)        0       1   
510       http://us.imdb.com/M/title-exact?Shichinin%20n...        0       1   
515       http://us.imdb.com/M/title-exact?Boot,%20Das%2...        0       1   
526       http://us.imdb.com/M/title-exact?Ben-Hur%20(1959)        0       1   
631       http://us.imdb.com/M/title-exact?Crying%20Game...        0       1   
651         http://us.imdb.com/M/title-exact?Glory%20(1989)        0       1   
686       http://us.imdb.com/M/title-exact?Perfect%20Wor...        0       1   
720       http://us.imdb.com/M/title-exact?First%20Knigh...        0       1   
806       http://us.imdb.com/M/title-exact?Menace%20II%2...        0       1   
840       http://us.imdb.com/M/title-exact?Last%20Man%20...        0       1   
855          http://us.imdb.com/M/title-exact?Diva%20(1981)        0       1   

          Adventure  Animation  Children  Comedy  Crime  ...  Horror  Musical   
movie_id                                                 ...                    
4                 0          0         0       1      0  ...       0        0  \
22                0          0         0       0      0  ...       0        0   
28                0          0         0       0      0  ...       0        0   
54                0          0         0       0      0  ...       0        0   
127               0          0         0       0      1  ...       0        0   
172        

In [39]:
acao_drama = filmes_com_mais_de_100_votos.query('Action == 1 and Drama == 1')
acao_drama.sort_values('Nota_Media', ascending = False).head(5)

movie_title release_date   
movie_id                                                 
127                 Godfather, The (1972)  01-Jan-1972  \
313                        Titanic (1997)  01-Jan-1997   
172       Empire Strikes Back, The (1980)  01-Jan-1980   
515                      Boot, Das (1981)  04-Apr-1997   
187        Godfather: Part II, The (1974)  01-Jan-1974   

                                                   IMDb_URL  unknown  Action   
movie_id                                                                       
127       http://us.imdb.com/M/title-exact?Godfather,%20...        0       1  \
313       http://us.imdb.com/M/title-exact?imdb-title-12...        0       1   
172       http://us.imdb.com/M/title-exact?Empire%20Stri...        0       1   
515       http://us.imdb.com/M/title-exact?Boot,%20Das%2...        0       1   
187       http://us.imdb.com/M/title-exact?Godfather:%20...        0       1   

          Adventure  Animation  Children  Comedy  Crime  ...  Horror  Musical   
movie_id                                                 ...                    
127               0          0         0       0      1  ...       0        0  \
313               0          0         0       0      0  ...       0        0   
172               1          0         0       0      0  ...       0        0   
515               0          0         0       0      0  ...       0        0   
187               0          0         0       0      1  ...       0        0   

          Mystery  Romance  SciFi  Thriller  War  Western  Total_de_Votos   
movie_id                                                                    
127             0        0      0         0    0        0             413  \
313             0        1      0         0    0        0             350   
172             0        1      1         0    1        0             367   
515             0        0      0         0    1        0             201   
187             0        0      0         0    0        0             209   

          Nota_Media  
movie_id              
127         4.283293  
313         4.245714  
172         4.204360  
515         4.203980  
187         4.186603  

[5 rows x 24 columns]

In [41]:
# Deste resultado, eu exclui os filmes que já assisti:
acao_drama.drop(assistido, errors = 'ignore').sort_values('Nota_Media', ascending = False).head(5)

movie_title release_date   
movie_id                                                 
127                 Godfather, The (1972)  01-Jan-1972  \
313                        Titanic (1997)  01-Jan-1997   
172       Empire Strikes Back, The (1980)  01-Jan-1980   
515                      Boot, Das (1981)  04-Apr-1997   
187        Godfather: Part II, The (1974)  01-Jan-1974   

                                                   IMDb_URL  unknown  Action   
movie_id                                                                       
127       http://us.imdb.com/M/title-exact?Godfather,%20...        0       1  \
313       http://us.imdb.com/M/title-exact?imdb-title-12...        0       1   
172       http://us.imdb.com/M/title-exact?Empire%20Stri...        0       1   
515       http://us.imdb.com/M/title-exact?Boot,%20Das%2...        0       1   
187       http://us.imdb.com/M/title-exact?Godfather:%20...        0       1   

          Adventure  Animation  Children  Comedy  Crime  ...  Horror  Musical   
movie_id                                                 ...                    
127               0          0         0       0      1  ...       0        0  \
313               0          0         0       0      0  ...       0        0   
172               1          0         0       0      0  ...       0        0   
515               0          0         0       0      0  ...       0        0   
187               0          0         0       0      1  ...       0        0   

          Mystery  Romance  SciFi  Thriller  War  Western  Total_de_Votos   
movie_id                                                                    
127             0        0      0         0    0        0             413  \
313             0        1      0         0    0        0             350   
172             0        1      1         0    1        0             367   
515             0        0      0         0    1        0             201   
187             0        0      0         0    0        0             209   

          Nota_Media  
movie_id              
127         4.283293  
313         4.245714  
172         4.204360  
515         4.203980  
187         4.186603  

[5 rows x 24 columns]

De acordo com o gênero escolhido e melhores média de avaliação, as recomendações são:

- O Poderoso Chefão;
- Titanic;
- Star Wars: O Império Contra-Ataca;
- O Barco;
- O Poderoso Chefão 2;

Mas, lembrando que, não considerei os filmes com menos de 100 avaliações. Isso faz com que filmes produzidos por pequenas produtoras ou que ainda não caíram no gosto do grande público, não sejam encontrados por novos usuários e desestimulando a procura por novas obras.

Fazendo um sistema de recomendação utilizando o histórico do usuário:

Eu compartilho o gosto de algumas coisas com meus amigos, por exemplo, gostamos de alguns filmes, séries e livros, ao mesmo tempo que não gostamos de outras obras. Ou seja, eu e meus amigos podemos avaliar essas obras com notas iguais ou próximas.

Para averiguar se compartilhamos os mesmos gostos, podemos calcular a distância euclidiana entre as avaliações, que em resumo, quanto menor for este valor, mais "compatível" e próximo será o gosto entre os perfis. Assim, se um usuário gostou de um filme que o outro perfil ainda não assistiu, será mais fácil efetuar recomendações.

Escolhendo o usuário 50

In [42]:
usuario_50 = notas.query('user_id == 50')
usuario_50[['item_id', 'rating']].set_index('item_id')

rating
item_id        
246           3
823           3
253           5
475           5
1084          5
286           2
9             4
125           2
123           4
325           1
508           5
288           4
319           5
324           5
276           2
1008          5
1010          5
268           4
544           4
15            2
327           3
124           1
547           4
100           2

In [43]:
def historico_usuario(usuario):
    historico_usuario = notas.query('user_id == %d' % usuario)
    historico_usuario = historico_usuario[['item_id', 'rating']].set_index('item_id')
    return historico_usuario

historico_usuario(50)

rating
item_id        
246           3
823           3
253           5
475           5
1084          5
286           2
9             4
125           2
123           4
325           1
508           5
288           4
319           5
324           5
276           2
1008          5
1010          5
268           4
544           4
15            2
327           3
124           1
547           4
100           2

Escolhendo outro usuário, agora o 75:

In [44]:
historico_usuario(75)

rating
item_id        
284           2
240           1
237           2
546           3
409           3
...         ...
222           5
866           2
129           3
864           4
473           3

[79 rows x 1 columns]

In [45]:
user50 = historico_usuario(43)
user75 = historico_usuario(62)

In [46]:
user50.join(user75, lsuffix='_do_user', rsuffix='_comparacao').dropna()

rating_do_user  rating_comparacao
item_id                                   
14                    2                4.0
815                   4                3.0
204                   4                3.0
151                   4                5.0
64                    5                4.0
...                 ...                ...
421                   3                5.0
127                   4                4.0
845                   5                3.0
4                     4                4.0
144                   4                3.0

[88 rows x 2 columns]

In [48]:
diferenca = user50.join(user75, lsuffix='_do_user',rsuffix='_comparacao').dropna()
np.linalg.norm(diferenca['rating_do_user'] - diferenca['rating_comparacao'])

13.564659966250536

Com esse valor, ainda não é possível dizermos se os perfis sã compatíveis. Vamos fazer uma função para simplificar o cálculo e seguir com a escolha de outros usuários.

In [49]:
def distancia_entre_perfis(user_id1, user_id2):
    notas1 = historico_usuario(user_id1)
    notas2 = historico_usuario(user_id2)
    diferenca = notas1.join(notas2, lsuffix = '_do_user', rsuffix = '_comparacao').dropna()
    distancia = np.linalg.norm(diferenca['rating_do_user'] - diferenca['rating_comparacao'])
    return [user_id1, user_id2, distancia]

In [50]:
distancia_entre_perfis(50, 75)

[50, 75, 3.4641016151377544]

In [51]:
notas.user_id.unique()

array([196, 186,  22, 244, 166, 298, 115, 253, 305,   6,  62, 286, 200,
       210, 224, 303, 122, 194, 291, 234, 119, 167, 299, 308,  95,  38,
       102,  63, 160,  50, 301, 225, 290,  97, 157, 181, 278, 276,   7,
        10, 284, 201, 287, 246, 242, 249,  99, 178, 251,  81, 260,  25,
        59,  72,  87,  42, 292,  20,  13, 138,  60,  57, 223, 189, 243,
        92, 241, 254, 293, 127, 222, 267,  11,   8, 162, 279, 145,  28,
       135,  32,  90, 216, 250, 271, 265, 198, 168, 110,  58, 237,  94,
       128,  44, 264,  41,  82, 262, 174,  43,  84, 269, 259,  85, 213,
       121,  49, 155,  68, 172,  19, 268,   5,  80,  66,  18,  26, 130,
       256,   1,  56,  15, 207, 232,  52, 161, 148, 125,  83, 272, 151,
        54,  16,  91, 294, 229,  36,  70,  14, 295, 233, 214, 192, 100,
       307, 297, 193, 113, 275, 219, 218, 123, 158, 302,  23, 296,  33,
       154,  77, 270, 187, 170, 101, 184, 112, 133, 215,  69, 104, 240,
       144, 191,  61, 142, 177, 203,  21, 197, 134, 180, 236, 26

In [52]:
print('O dataset possui %d usuarios.' % len(notas.user_id.unique()))

O dataset possui 943 usuarios.


No df, temos no total 943 usuários, portanto vamos compararar avaliação do usuário 50 com outros 5 usuários.

In [53]:
usuario1 = 50

distancias = []
for usuario in notas['user_id'].unique():
    calculo = distancia_entre_perfis(usuario1, usuario)
    distancias.append(calculo)

distancias[:5]

[[50, 196, 3.0],
 [50, 186, 3.605551275463989],
 [50, 22, 0.0],
 [50, 244, 4.69041575982343],
 [50, 166, 1.4142135623730951]]

Vimos que o resultado mostrou a distância entre o usuário 50 com os demais pode ser menor.

In [54]:
# Função para calcular a distância entre 1 usuário específico e os demais:

def distancia_entre_usuarios(usuario_1):
    todos_os_usuarios = notas['user_id'].unique()
    distancias = [distancia_entre_perfis(usuario_1, user_id) for user_id in todos_os_usuarios]
    distancias = pd.DataFrame(distancias, columns = ['Usuario_1', 'Outro_user', 'Distancia'])
    return distancias

distancia_entre_usuarios(50).head()

Usuario_1  Outro_user  Distancia
0         50         196   3.000000
1         50         186   3.605551
2         50          22   0.000000
3         50         244   4.690416
4         50         166   1.414214

In [55]:
# Ordenando o Dataframe pela Distância:

def mais_proximos_de(usuario_1):
    distancias = distancia_entre_usuarios(usuario_1)
    distancias = distancias.sort_values('Distancia')
    distancias = distancias.set_index('Outro_user').drop(usuario_1)
    return distancias

mais_proximos_de(50)

Usuario_1  Distancia
Outro_user                      
746                50   0.000000
904                50   0.000000
564                50   0.000000
565                50   0.000000
726                50   0.000000
...               ...        ...
104                50   8.717798
342                50   8.944272
592                50   9.165151
445                50   9.848858
181                50  10.535654

[942 rows x 2 columns]

In [56]:
mais_proximos_de(50).head(20)

Usuario_1  Distancia
Outro_user                      
746                50        0.0
904                50        0.0
564                50        0.0
565                50        0.0
726                50        0.0
738                50        0.0
245                50        0.0
162                50        0.0
8                  50        0.0
208                50        0.0
217                50        0.0
272                50        0.0
911                50        0.0
691                50        0.0
415                50        0.0
914                50        0.0
33                 50        0.0
584                50        0.0
918                50        0.0
515                50        0.0

A partir do gabarito: gerando recomendações com KNN.

KNN significa *K-nearest neighbors*, que traduzindo é *K-vizinhos mais próximos*. Este método é parecido com o que foi feito até agora, quando buscamos os perfis de usuários que fossem mais próximos, através dos filmes que eles assistiram e notas que atribuíram.

Nesta fase, reformulei algumas funções. Por exemplo, a função **distancia_entre_perfis** recebeu um novo argumento, caso a quantidade de filmes que ambos assistiram for menor do que 5, a distância entre estes perfis não é calculada. Na função **distancia_entre_usuarios**, o novo argumento determina a quantidade máxima de perfis que se quer comparar. Já na função **mais_proximos_de**, adicionei os 2 argumentos citados anteriormente.

In [57]:
def distancia_entre_perfis(user_id1, user_id2, minimo = 5):
    notas1 = historico_usuario(user_id1)
    notas2 = historico_usuario(user_id2)
    diferenca = notas1.join(notas2, lsuffix = '_do_user', rsuffix = '_comparacao').dropna()
    
    if(len(diferenca) < minimo):
        return None
    
    distancia = np.linalg.norm(diferenca['rating_do_user'] - diferenca['rating_comparacao'])
    return [user_id1, user_id2, distancia]

In [58]:
def distancia_entre_usuarios(user_1, numero_maximo_de_analise = None):
    todos_os_usuarios = notas['user_id'].unique()
    
    if numero_maximo_de_analise:
        todos_os_usuarios = todos_os_usuarios[:numero_maximo_de_analise]
        
    distancias = [distancia_entre_perfis(user_1, user_id) for user_id in todos_os_usuarios]
    distancias = list(filter(None, distancias))
    distancias = pd.DataFrame(distancias, columns = ['Usuario_1', 'Outro_user', 'Distancia'])
    return distancias

In [59]:
def mais_proximos_de(user_1, quantidade_user_proximos = 10, numero_maximo_de_analise = None):
    distancias = distancia_entre_usuarios(user_1, numero_maximo_de_analise = numero_maximo_de_analise)
    distancias = distancias.sort_values('Distancia')
    distancias = distancias.set_index('Outro_user').drop(user_1)
    return distancias.head(quantidade_user_proximos)

In [60]:
mais_proximos_de(196, numero_maximo_de_analise = 50)

Usuario_1  Distancia
Outro_user                      
251               196   1.732051
97                196   2.236068
81                196   2.449490
157               196   2.645751
287               196   3.162278
160               196   3.162278
298               196   3.162278
119               196   3.162278
290               196   3.316625
10                196   3.464102

Ao testar a última função, vemos que os 10 perfis (dentro dos 50 primeiros presentes no dataset *notas*) os que tiveram maior proximidade com o usuário 196, são os usuários 251 e 97.

---

Tendo conhecimento dos perfis mais próximos, agora é possível buscar novos filmes para o usuário.

Primeiro, criei uma nova função que possui os mesmos argumentos que a função **mais_proximos_de**, que, em resumo, retorna 5 recomendações de filmes, ao fazer uma seleção daqueles que tiveram as melhores notas nos perfis mais próximos.

In [62]:
def sugestoes(user1, quantidade_user_proximos = 10, numero_maximo_de_analise = None):
    notas_user1 = historico_usuario(user1)
    historico_filmes = notas_user1.index
    similares = mais_proximos_de(user1, quantidade_user_proximos = quantidade_user_proximos, 
                                 numero_maximo_de_analise = numero_maximo_de_analise)
    usuarios_similares = similares.index
    notas_dos_similares = notas.set_index('user_id').loc[usuarios_similares]
    recomendacoes = notas_dos_similares.groupby('item_id').mean()[['rating']]
    recomendacoes = recomendacoes.sort_values('rating', ascending = False)
    return recomendacoes.join(data_filmes).head()

sugestoes(196, quantidade_user_proximos = 2, numero_maximo_de_analise = 50)

rating                                      movie_title release_date   
item_id                                                                         
237         5.0                             Jerry Maguire (1996)  13-Dec-1996  \
612         5.0                              Lost Horizon (1937)  01-Jan-1937   
115         5.0  Haunted World of Edward D. Wood Jr., The (1995)  26-Apr-1996   
482         5.0                          Some Like It Hot (1959)  01-Jan-1959   
480         5.0                        North by Northwest (1959)  01-Jan-1959   

                                                  IMDb_URL  unknown  Action   
item_id                                                                       
237      http://us.imdb.com/M/title-exact?Jerry%20Magui...        0       0  \
612      http://us.imdb.com/M/title-exact?Lost%20Horizo...        0       0   
115      http://us.imdb.com/Title?Haunted+World+of+Edwa...        0       0   
482      http://us.imdb.com/M/title-exact?Some%20Like%2...        0       0   
480      http://us.imdb.com/M/title-exact?North%20by%20...        0       0   

         Adventure  Animation  Children  Comedy  ...  Horror  Musical   
item_id                                          ...                    
237              0          0         0       0  ...       0        0  \
612              0          0         0       0  ...       0        0   
115              0          0         0       0  ...       0        0   
482              0          0         0       1  ...       0        0   
480              0          0         0       1  ...       0        0   

         Mystery  Romance  SciFi  Thriller  War  Western  Total_de_Votos   
item_id                                                                    
237            0        1      0         0    0        0             384  \
612            0        0      0         0    0        0              34   
115            0        0      0         0    0        0              15   
482            0        0      0         0    0        0             128   
480            0        0      0         1    0        0             179   

         Nota_Media  
item_id              
237        3.710938  
612        3.941176  
115        3.733333  
482        3.984375  
480        4.284916  

[5 rows x 25 columns]

As indicações (usando apenas 2 usuários) são:
- Jerry Maguire: A Grande Virada;
- Horizonte Perdido;
- The Haunted World of Edward D. Wood Jr.;
- Quanto Mais Quente Melhor;
- Intriga Internacional.

In [63]:
sugestoes(196, numero_maximo_de_analise = 50)

rating                                      movie_title release_date   
item_id                                                                         
508         5.0               People vs. Larry Flynt, The (1996)  27-Dec-1996  \
531         5.0                                     Shine (1996)  22-Nov-1996   
115         5.0  Haunted World of Edward D. Wood Jr., The (1995)  26-Apr-1996   
272         5.0                         Good Will Hunting (1997)  01-Jan-1997   
506         5.0                     Rebel Without a Cause (1955)  01-Jan-1955   

                                                  IMDb_URL  unknown  Action   
item_id                                                                       
508      http://us.imdb.com/M/title-exact?People%20vs.%...        0       0  \
531        http://us.imdb.com/M/title-exact?Shine%20(1996)        0       0   
115      http://us.imdb.com/Title?Haunted+World+of+Edwa...        0       0   
272      http://us.imdb.com/M/title-exact?imdb-title-11...        0       0   
506      http://us.imdb.com/M/title-exact?Rebel%20Witho...        0       0   

         Adventure  Animation  Children  Comedy  ...  Horror  Musical   
item_id                                          ...                    
508              0          0         0       0  ...       0        0  \
531              0          0         0       0  ...       0        0   
115              0          0         0       0  ...       0        0   
272              0          0         0       0  ...       0        0   
506              0          0         0       0  ...       0        0   

         Mystery  Romance  SciFi  Thriller  War  Western  Total_de_Votos   
item_id                                                                    
508            0        0      0         0    0        0             215  \
531            0        1      0         0    0        0             129   
115            0        0      0         0    0        0              15   
272            0        0      0         0    0        0             198   
506            0        0      0         0    0        0              90   

         Nota_Media  
item_id              
508        3.595349  
531        3.922481  
115        3.733333  
272        4.262626  
506        3.855556  

[5 rows x 25 columns]

Mas ao utilizar os perfis dos 10 usuários mais próximos, temos:
- O Povo Contra Larry Flynt;
- Shine - Brilhante;
- The Haunted World of Edward D. Wood Jr.;
- Gênio Indomável;
- Juventude Transviada.

In [64]:
sugestoes(196)

rating                   movie_title release_date   
item_id                                                      
185         5.0                 Psycho (1960)  01-Jan-1960  \
135         5.0  2001: A Space Odyssey (1968)  01-Jan-1968   
582         5.0             Piano, The (1993)  01-Jan-1993   
583         5.0      Romeo Is Bleeding (1993)  01-Jan-1993   
488         5.0           Sunset Blvd. (1950)  01-Jan-1950   

                                                  IMDb_URL  unknown  Action   
item_id                                                                       
185       http://us.imdb.com/M/title-exact?Psycho%20(1960)        0       0  \
135      http://us.imdb.com/M/title-exact?2001:%20A%20S...        0       0   
582      http://us.imdb.com/M/title-exact?Piano,%20The%...        0       0   
583      http://us.imdb.com/M/title-exact?Romeo%20Is%20...        0       0   
488      http://us.imdb.com/M/title-exact?Sunset%20Boul...        0       0   

         Adventure  Animation  Children  Comedy  ...  Horror  Musical   
item_id                                          ...                    
185              0          0         0       0  ...       1        0  \
135              0          0         0       0  ...       0        0   
582              0          0         0       0  ...       0        0   
583              0          0         0       0  ...       0        0   
488              0          0         0       0  ...       0        0   

         Mystery  Romance  SciFi  Thriller  War  Western  Total_de_Votos   
item_id                                                                    
185            0        1      0         1    0        0             239  \
135            1        0      1         1    0        0             259   
582            0        1      0         0    0        0             168   
583            0        0      0         1    0        0              37   
488            0        0      0         0    0        0              65   

         Nota_Media  
item_id              
185        4.100418  
135        3.969112  
582        3.601190  
583        3.000000  
488        4.200000  

[5 rows x 25 columns]

E quando não definimos um número máximo de usuários para análise, temos:
- Psicose;
- 2001 - Uma Odisseia no Espaço;
- O Piano;
- O Sangue de Romeo;
- Crepúsculo dos Deuses.

# Gerando recomendação para um novo usuário:

Vamos considerar os filmes do usuário 11

In [65]:
data_filmes.sample(11)

movie_title release_date   
movie_id                                                                   
1318                                         Catwalk (1995)  07-Jun-1996  \
603                                      Rear Window (1954)  01-Jan-1954   
1531      Far From Home: The Adventures of Yellow Dog (1...  01-Jan-1995   
681                                       Wishmaster (1997)  01-Jan-1997   
540                                      Money Train (1995)  01-Jan-1995   
66                           While You Were Sleeping (1995)  01-Jan-1995   
1281                                      Manny & Lo (1996)  26-Jul-1996   
273                                             Heat (1995)  01-Jan-1995   
989                                 Cats Don't Dance (1997)  26-Mar-1997   
1389                                           Mondo (1996)  27-Jun-1997   
876                                      Money Talks (1997)  22-Aug-1997   

                                                   IMDb_URL  unknown  Action   
movie_id                                                                       
1318              http://us.imdb.com/Title?Catwalk+(1995/I)        0       0  \
603       http://us.imdb.com/M/title-exact?Rear%20Window...        0       0   
1531      http://us.imdb.com/M/title-exact?Far%20From%20...        0       0   
681       http://us.imdb.com/M/title-exact?Wishmaster+(1...        0       0   
540       http://us.imdb.com/M/title-exact?Money%20Train...        0       1   
66        http://us.imdb.com/M/title-exact?While%20You%2...        0       0   
1281      http://us.imdb.com/M/title-exact?Manny%20&%20L...        0       0   
273          http://us.imdb.com/M/title-exact?Heat%20(1995)        0       1   
989       http://us.imdb.com/M/title-exact?Cats%20Don%27...        0       0   
1389          http://us.imdb.com/M/title-exact?Mondo+(1996)        0       0   
876       http://us.imdb.com/M/title-exact?Money+Talks+(...        0       1   

          Adventure  Animation  Children  Comedy  Crime  ...  Horror  Musical   
movie_id                                                 ...                    
1318              0          0         0       0      0  ...       0        0  \
603               0          0         0       0      0  ...       0        0   
1531              1          0         1       0      0  ...       0        0   
681               0          0         0       0      0  ...       1        0   
540               0          0         0       0      0  ...       0        0   
66                0          0         0       1      0  ...       0        0   
1281              0          0         0       0      0  ...       0        0   
273               0          0         0       0      1  ...       0        0   
989               0          1         1       0      0  ...       0        1   
1389              0          0         0       0      0  ...       0        0   
876               0          0         0       1      0  ...       0        0   

          Mystery  Romance  SciFi  Thriller  War  Western  Total_de_Votos   
movie_id                                                                    
1318            0        0      0         0    0        0               3  \
603             1        0      0         1    0        0             209   
1531            0        0      0         0    0        0               7   
681             0        0      0         0    0        0              27   
540             0        0      0         0    0        0              43   
66              0        1      0         0    0        0             162   
1281            0        0      0         0    0        0              13   
273             0        0      0         1    0        0             223   
989             0        0      0         0    0        0              32   
1389            0        0      0         0    0        0               3   
876             0        0      0       

E para adicioná-lo ao nosso sistema, fiz uma função que irá atribuir um número de ID e registrar os filmes e as notas. Escolhi as notas deste novo usuário de forma randômica:

In [66]:
import random
notas_novo_user = [random.randint(1,5) for x in range(11)]
notas_novo_user

[2, 5, 3, 1, 1, 2, 5, 4, 4, 5, 5]

In [67]:
def novo_usuario(seus_filmes):
    novo_usuario = notas['user_id'].max()+1
    notas_do_usuario_novo = pd.DataFrame(seus_filmes, columns = ['item_id', 'rating'])
    notas_do_usuario_novo['user_id'] = novo_usuario
    return pd.concat([notas, notas_do_usuario_novo])

In [68]:
notas = novo_usuario([[723, 3], [1094, 3], [652, 1], [1522, 5], [1243, 2], [1456, 1], [1086, 5], [1034, 2], [703, 2], [90, 3], [217, 2]])
notas.tail(12)

user_id  item_id  rating    timestamp
99999       12      203       3  879959583.0
0          944      723       3          NaN
1          944     1094       3          NaN
2          944      652       1          NaN
3          944     1522       5          NaN
4          944     1243       2          NaN
5          944     1456       1          NaN
6          944     1086       5          NaN
7          944     1034       2          NaN
8          944      703       2          NaN
9          944       90       3          NaN
10         944      217       2          NaN

In [69]:
# Para o novo usuário temos as seguintes recomendações:

sugestoes(944)

rating                       movie_title release_date   
item_id                                                          
506         5.0      Rebel Without a Cause (1955)  01-Jan-1955  \
64          5.0  Shawshank Redemption, The (1994)  01-Jan-1994   
475         5.0              Trainspotting (1996)  19-Jul-1996   
156         5.0             Reservoir Dogs (1992)  01-Jan-1992   
50          5.0                  Star Wars (1977)  01-Jan-1977   

                                                  IMDb_URL  unknown  Action   
item_id                                                                       
506      http://us.imdb.com/M/title-exact?Rebel%20Witho...        0       0  \
64       http://us.imdb.com/M/title-exact?Shawshank%20R...        0       0   
475          http://us.imdb.com/Title?Trainspotting+(1996)        0       0   
156      http://us.imdb.com/M/title-exact?Reservoir%20D...        0       0   
50       http://us.imdb.com/M/title-exact?Star%20Wars%2...        0       1   

         Adventure  Animation  Children  Comedy  ...  Horror  Musical   
item_id                                          ...                    
506              0          0         0       0  ...       0        0  \
64               0          0         0       0  ...       0        0   
475              0          0         0       0  ...       0        0   
156              0          0         0       0  ...       0        0   
50               1          0         0       0  ...       0        0   

         Mystery  Romance  SciFi  Thriller  War  Western  Total_de_Votos   
item_id                                                                    
506            0        0      0         0    0        0              90  \
64             0        0      0         0    0        0             283   
475            0        0      0         0    0        0             250   
156            0        0      0         1    0        0             148   
50             0        1      1         0    1        0             583   

         Nota_Media  
item_id              
506        3.855556  
64         4.445230  
475        3.884000  
156        3.986486  
50         4.358491  

[5 rows x 25 columns]

- Juventude Transviada;
- Um Sonho de Liberdade;
- Trainspotting - Sem Limites;
- Cães de Aluguel;
- Star Wars: Episódio IV – Uma Nova Esperança.